dynamiqs/propagator/sepropagator.py

```diff
@@ -0,0 +1,90 @@
+from __future__ import annotations
+
+import jax.numpy as jnp
+from jaxtyping import ArrayLike
+
+from ..core.expm_solver import ExpmSolver
+from ..gradient import Gradient
+from ..options import Options
+from ..result import PropagatorResult, Saved
+from ..sesolve import sesolve
+from ..solver import Expm, Solver, Tsit5
+from ..time_array import ConstantTimeArray, PWCTimeArray, TimeArray
+from ..utils.operators import eye
+
+__all__ = ["propagator"]
+
+
+def propagator(
+    H: ArrayLike | TimeArray,
+    tsave: ArrayLike,
+    *,
+    solver: Solver = None,
+    gradient: Gradient | None = None,
+    options: Options = Options(),  # noqa: B008
+) -> PropagatorResult:
+    r"""Solve the Schrödinger equation by computing the propagator.
+
+    This computation is done in one of two ways. If `solver` is set to `None`
+    and if the Hamiltonian is of type [`dq.ConstantTimeArray`][dynamiqs.ConstantTimeArray] or
+    [`dq.PWCTimeArray`][dynamiqs.PWCTimeArray], then the propagator is computed
+    by appropriately exponentiating the Hamiltonian. On the other hand if the solver is
+    specified or if the Hamiltonian is not constant or piece-wise constant, we solve
+    the Schrödinger equation using [`dq.sesolve`][dynamiqs.sesolve] and batching over
+    all initial states.
+
+    Args:
+        H _(array-like or time-array of shape (...H, n, n))_: Hamiltonian.
+        tsave _(array-like of shape (ntsave,))_: Times at which the propagators
+            are saved. The equation is solved from `tsave[0]` to
+            `tsave[-1]`, or from `t0` to `tsave[-1]` if `t0` is specified in `options`.
+        solver: Solver for the integration. As mentioned above, if this is `None`
+            and the Hamiltonian is constant or piece-wise constant, the propagator is computed
+            by matrix exponential. If it is `None` and the Hamiltonian is not
+            constant or piece-wise constant, the propagator is computed using
+            [`dq.sesolve`][dynamiqs.sesolve] with the solver defaulting to
+            [`dq.solver.Tsit5`][dynamiqs.solver.Tsit5]. If the solver is specified, the propagator
+            is computed using [`dq.sesolve`][dynamiqs.sesolve]. See [`dq.sesolve`][dynamiqs.sesolve]
+            for a list of supported solvers.
+
+        gradient: Algorithm used to compute the gradient.
+        options: Generic options, see [`dq.Options`][dynamiqs.Options].
+
+    Returns:
+        [`dq.PropagatorResult`][dynamiqs.PropagatorResult] object holding the result of the
+            propagator computation. Use the attribute `propagator`
+            to access saved quantities, more details in
+            [`dq.PropagatorResult`][dynamiqs.PropagatorResult].
+    """  # noqa: E501
+    if not options.cartesian_batching:
+        raise ValueError(
+            "Flat batching not supported for propagator. "
+            "Only cartesian batching is supported"
+        )
```

Comment:
Maybe, though I wonder about the behavior in the case where we default to `sesolve`? Like I can imagine a case where the Hamiltonian has dimension (n, n, n) and then `sesolve` gets confused and tries to apply the first dimension of H to the first dimension of initial states which has shape (n, n, 1). This would be undesired.

Reply:
Yes, but I guess we should not be passing this option down to `sesolve` then. This is an option related to `propagator` and should be transparent for users (even if we have to make dirty tricks in the backend regarding how we call `sesolve`).

Reply:
Gotcha: see newest version of the code for a hack to set `cartesian_batching` to `True` if we are using `sesolve`